### ISSUE TITLE
Pipx package in Ubuntu 24.04 is obsolete, how to properly upgrade to 1.6.0

### ISSUE THREAD

**Issue by FredM67** [pypa/pipx#1481]
Pipx is in version 1.4.3 in Ubuntu 24-04.

How to upgrade to the 1.6.0 properly ? (without breaking anything ! )

**Comment by huxuan** [comment]
I personally recommend install pipx with pipx!

```bash
pipx install pipx
# And optional remove the obsolete apt pipx!
sudo apt purge --autoremove pipx
```

You may need to restart the shell or run command `hash -r` to make `pipx` command work.

Reference: https://github.com/pypa/pipx/discussions/1341 https://github.com/pypa/pipx/issues/1449

**Comment by smanleyjr-alg** [comment]
On ubuntu 24.04
this puts **pipx** in **~/.local/bin** instead of **/usr/local/bin** 
``` sudo apt update 
sudo apt install pipx               # installs pipx version 1.4.3 to /usr/local/bin  recall --global is busted
pipx ensurepath                     # adds ~/.local/bin to path
pipx install pipx                   # installs latest pipx to ~/.local/bin 
sudo apt purge --autoremove pipx    # remove pipx 1.4.3 via apt,  and all extra files including autocomplete from /usr/local/bin
``` 
restart shell then try pipx from sudo
results in this failing: 
```
sudo pipx ensurepath --global       # <-- fails, pipx command not found
``` 
To Fix I continued with:
``` 
sudo ~/.local/bin/pipx install pipx --global   #installs latest pipx version to /usr/local/bin because --global works.
pipx uninstall pipx                            # remove the local version
``` 
restart the shell
``` 
pipx ensurepath                                # make sure the ~/.local/bin is added to $PATH
sudo pipx ensurepath --global                  # make sure that /usr/local/bin is set to global install 
``` 
It all works far as I can tell. No issues with using pipx or upgrading it. Net effect is a way of installing/updating without editing a environment var, path, or rc file by hand...

**Comment by zinc75** [comment]
Hi everyone,

I’ve created a backported version of **pipx 1.6.0** for **Ubuntu 24.04 LTS** (Noble), addressing this issue and thos like [[#1524](https://github.com/pypa/pipx/issues/1524)](https://github.com/pypa/pipx/issues/1524) where `sudo pipx ensurepath --global` fails.

You can find the backport (from official Ubuntu 24.10 sources) along with the compiled `.deb` package and build instructions in my repository:

[https://github.com/zinc75/pipx-1.6.0-backport-ubuntu-2404-lts](https://github.com/zinc75/pipx-1.6.0-backport-ubuntu-2404-lts)

### What’s included:
- Precompiled `.deb` package for Ubuntu 24.04.
- Source files, build scripts, and method to compile using sources from Ubuntu 24.10.
  
Installation is straightforward with the `.deb` package, or you can build from source if preferred. 

Enjoy !

**Comment by sofigio** [comment]
@zinc75 nice but what's the point in cloning the repository? we can just `wget` the binary from github... or will you keep it updated with newer versions? In that case an apt repository would be perfect. 

For example I just want to install glances, I'm not a python developer, and I'm looking for a solution as simple as possible, to replicate it on all my new machines. (For ubuntu <24 I just did `apt install python3-pip; pip3 install glances`

I don't like either `pipx ensurepath` because adding `/root/.local/bin` to the path seems wrong... 

This is the best solution I found so far (all done as root):

```bash
apt install pipx # Install 1.4
pipx install pipx # Install 1.7 in ~/.local/bin/
apt purge --autoremove pipx # Remove 1.4
~/.local/bin/pipx install --global glances # Install 1.7 in /usr/local/bin/pipx
pipx remove pipx # Remove 1.7 from ~/.local/

ll ~/.local/bin/ # Make sure it's empty
whereis pipx # will give pipx: /usr/local/bin/pipx
```

Now pipx will be available for all users. 
For example you can 
`sudo pipx --global install glances`

**Comment by zinc75** [comment]
@sofigio : I agree that getting the .deb is the easiest way. That's the exact reason why I made a release in my github repo with the .deb ! ( [pipx_1.6.0-1_all.deb](https://github.com/zinc75/pipx-1.6.0-backport-ubuntu-2404-lts/releases/download/1.6.0-1/pipx_1.6.0-1_all.deb) .

Installing it is easy as downloading it and installing it using `sudo dpkg -i pipx_1.6.0-1_all.deb`

The whole point of making it available on a repo is for people who would like to check I did not modify anything to the sources (for obvious security reasons), and to document the method of backporting a .deb, for reproducibility. That's all the point of open sourcing, isn't it ? 

From your proposed solution, you install pipx using pipx (two times), and the official documentation (https://github.com/pypa/pipx#install-pipx)  states : 

> It is not recommended to install pipx via pipx. If you'd like to do this anyway, take a look at the [pipx-in-pipx](https://github.com/mattsb42-meta/pipx-in-pipx) project and read about the limitations there. 

That's why I decided to make a backport of the Ubuntu 24.10 version. Choice is everywhere, and the repo is there both for advanced users and end users. There is no obligation to use it ;-) 

Best,

**Comment by zinc75** [comment]
@sofigio : 


> This is the best solution I found so far (**all done as root**):

Doing anything as root is, as far as I know, not the way to go with Ubuntu ... 

> I don't like `pipx ensurepath` because adding `/root/.local/bin` to the path seems wrong...

This does not (when not doing it as root !)

`pipx ensurepath`   makes sure the ~/.local/bin is added to $PATH, and `sudo pipx ensurepath --global` makes sure that `/usr/local/bin` is set to global install. 

 Best regards,

**Comment by martimors** [comment]
This causes `pipx` to be unusable on `ansible`. https://docs.ansible.com/ansible/latest/collections/community/general/pipx_module.html

**Comment by rursprung** [comment]
i guess it might make sense to report this to the downstream ubuntu packaging rather than reporting it here (though this issue did help me to find the answer on how to use `pipx` to update itself since i needed `--global` 🙂): https://launchpad.net/ubuntu/+source/python-pipx/+bugs

based on [this documentation](https://documentation.ubuntu.com/sru/en/latest/) i'm not 100% sure if they can just update the `pipx` version as it talks about freezing dependency versions. in newer ubuntu versions they also packaged newer `pipx` versions

**Comment by sagisar1** [comment]
@rursprung I agree with you, this is a huge problem, because since the transition to system-wide packages (which I bless and encourage), still not all python modules are in the repositories, and therefore you need pipx, but without the --global option you can't run it under all users.
Did you reported it to ubuntu packaging already?

Anyway for anyone who reads this, for now my solution is using the pipx.pyz file

**Comment by rursprung** [comment]
> Did you reported it to ubuntu packaging already?

no, i don't have an account there (i already collect enough accounts as it is; don't need one more just for that ticket)

**Comment by ilpianista** [comment]
> Did you reported it to ubuntu packaging already?

Done. https://bugs.launchpad.net/ubuntu/+source/python-pipx/+bug/2131679

**Comment by tim-oe** [comment]
with trixie based debian OSes. you can at least upgrade to the latest

**Comment by Chris2048** [comment]
I found that `apt install pipx` did not work for me, apt on Ubuntu Noble cannot find the package,
**but** I found the package in the ubuntu package archive, and you can install relevant package from there to bootstrap;

Here's a rough script for doing so.
_run in bash shell, inspecting line-by-line so you understand what's going on._

```
mkdir /tmp/install_pipx/ &&
pushd /tmp/install_pipx/ &&
wget https://launchpad.net/ubuntu/+archive/primary/+files/pipx_1.4.3-1_all.deb &&
wget https://launchpad.net/ubuntu/+archive/primary/+files/python3-userpath_1.9.1-1_all.deb &&
mkdir tmp_install/ &&
dpkg-deb -x python3-userpath_1.9.1-1_all.deb ./tmp_install/ &&
dpkg-deb -x pipx_1.4.3-1_all.deb ./tmp_install/ &&
PYTHONPATH="./tmp_install/usr/lib/python3/dist-packages/" ./tmp_install/usr/bin/pipx install --force pipx &&
pipx ensurepath &&  eval "$(register-python-argcomplete pipx)";
popd && rm -rf /tmp/install_pipx/
```

Now, `which pipx` should show the local install.

**Comment by tim-oe** [comment]
on ubuntu 22 the following worked

# upgrad pip using pip
python3 -m pip install --upgrade pip
# install pipx with upgrade pip
python3 -m pip install --upgrade pipx

**Comment by Dzeri96** [comment]
> This causes `pipx` to be unusable on `ansible`. https://docs.ansible.com/ansible/latest/collections/community/general/pipx_module.html

I know this might not be the best place, but here's a fix for anyone who's struggling with Ansible and waiting for a new pipx version to be available in their distro's repo:
```yaml
# TODO: Install pipx over APT once it reaches v1.7
- name: Check if pipx is installed locally
  ansible.builtin.stat:
    path: "{{ ansible_facts['user_dir'] }}/.local/bin/pipx"
  register: pipx_stat

- name: Bootstrap pipx
  ansible.builtin.shell: |
    python3 -m venv /tmp/bootstrap
    /tmp/bootstrap/bin/pip install pipx
    /tmp/bootstrap/bin/pipx install pipx
    rm -rf /tmp/bootstrap
    .local/bin/pipx ensurepath
  when: not pipx_stat.stat.exists

- name: Install packages via pipx
  vars:
    ansible_python_interpreter: "{{ ansible_facts['user_dir']}}/.local/share/pipx/venvs/pipx/bin/python"
  community.general.pipx:
    name: "{{ item }}"
    install_deps: true
  loop:
    - uv
    - ansible
```
I got this idea from https://pipx.pypa.io/stable/how-to/install-pipx/#self-managed-pipx, however, the last step is wrong in the documentation. I'll have to file an issue for that.

I don't like the usage of `ansible_python_interpreter`, but it is what it is.
